library/bigip_asm_policy.py

```diff
@@ -0,0 +1,546 @@
+      user: "admin"
+      password: "admin"
+      name: "active_policy"
+      state: "present"
+  delegate_to: localhost
+  
+'''
+
+RETURN = '''
+active:
+    description: Set when activating/deactivating ASM policy
+    returned: changed
+    type: bool
+    sample: yes
+state:
+    description: Action performed on the target device.
+    returned: changed
+    type: string
+    sample: "absent"
+file:
+    description: Local path to ASM policy XML file.
+    returned: changed
+    type: string
+    sample: "/root/some_policy.xml"
+template:
+    description: Name of the built-in ASM policy template
+    returned: changed
+    type: string
+    sample: "POLICY_TEMPLATE_SHAREPOINT_2007_HTTP"
+name:
+    description: Name of the ASM policy to be managed/created
+    returned: changed
+    type: string
+    sample: "Asm_APP1_Transparent"
+'''
+
+
+
+import os
+import time
+from icontrol.exceptions import iControlUnexpectedHTTPError
+from ansible.module_utils.f5_utils import (
+    AnsibleF5Client,
+    AnsibleF5Parameters,
+    defaultdict,
+    HAS_F5SDK,
+    F5ModuleError,
+    iteritems,
+)
+
+
+class Parameters(AnsibleF5Parameters):
+    def __init__(self, params=None):
+        self._values = defaultdict(lambda: None)
+        if params:
+            self.update(params=params)
+
+    def update(self, params=None):
+        if params:
+            for k, v in iteritems(params):
+                if self.api_map is not None and k in self.api_map:
+                    map_key = self.api_map[k]
+                else:
+                    map_key = k
+
+                # Handle weird API parameters like `dns.proxy.__iter__` by
+                # using a map provided by the module developer
+                class_attr = getattr(type(self), map_key, None)
+                if isinstance(class_attr, property):
+                    # There is a mapped value for the api_map key
+                    if class_attr.fset is None:
+                        # If the mapped value does not have
+                        # an associated setter
+                        self._values[map_key] = v
+                    else:
+                        # The mapped value has a setter
+                        setattr(self, map_key, v)
+                else:
+                    # If the mapped value is not a @property
+                    self._values[map_key] = v
+
+    updatables = [
+        'active'
+    ]
+
+    returnables = [
+        'name', 'template', 'file', 'active'
+    ]
+
+    api_attributes = [
+        'name', 'file'
+    ]
+    api_map = {
+        'filename': 'file'
+    }
+
+    @property
+    def template(self):
+        tmpl_name = self._values['template']
+        if tmpl_name is None:
+            return None
+        if self._template_exists_on_device(tmpl_name):
+            return self._values['template']
+        else:
+            raise F5ModuleError('Template with the given name: {0} does not exist'.format(tmpl_name))
+
```

Comment:
i generally do not attempt to check to see if something already exists like this because MCPD will do this for you. You're either making a round-trip to bigip here, or you're making a round-trip to bigip during the update (when mcpd will catch your error). Therefore I delegate that to mcpd because I am not the expert.

Reply:
We need to be able to check for this so the create on device operstion does not fail